aiida/storage/psql_dos/migrator.py

```diff
@@ -148,47 +148,164 @@ def validate_storage(self) -> None:
+
+        initialised: bool = False
+
+        if not self.is_initialised:
+            self.initialise_repository()
+            self.initialise_database()
+            initialised = True
+
+        # Call migrate in the case the storage was already initialised but not yet at the latest schema version. If it
+        # was, then the following is a no-op anyway.
+        self.migrate()
+
+        return initialised
+
+    @property
+    def is_initialised(self) -> bool:
+        """Return whether the storage is initialised.
+
+        This is the case if both the database and the repository are initialised.
+
+        :returns: ``True`` if the storage is initialised, ``False`` otherwise.
+        """
+        return self.is_repository_initialised and self.is_database_initialised
+
+    @property
+    def is_repository_initialised(self) -> bool:
+        """Return whether the repository is initialised.
+
+        :returns: ``True`` if the repository is initialised, ``False`` otherwise.
+        """
+        return self.get_container().is_initialised
+
+    @property
+    def is_database_initialised(self) -> bool:
+        """Return whether the database is initialised.
+
+        This is the case if it contains the table that holds the schema version for alembic or Django.
+
+        :returns: ``True`` if the database is initialised, ``False`` otherwise.
+        """
+        with self._connection_context() as connection:
+            return (
+                inspect(connection).has_table(self.alembic_version_tbl_name) or
+                inspect(connection).has_table(self.django_version_table.name)
+            )
+
+    def reset_repository(self) -> None:
+        """Reset the repository by deleting all of its contents.
+
+        This will also destroy the configuration and so in order to use it again, it will have to be reinitialised.
+        """
+        import shutil
+        try:
+            shutil.rmtree(self.get_container().get_folder())
+        except FileNotFoundError:
+            pass
+
+    def reset_database(self) -> None:
+        """Reset the database by deleting all content from all tables except for the alembic version table.
+
+        This will also destroy the settings table and so in order to use it again, it will have to be reinitialised.
+        """
+        self.delete_tables(exclude_tables=[self.alembic_version_tbl_name])
+
+    def delete_tables(self, exclude_tables: list[str] | None = None) -> None:
+        """Delete the tables of the database.
+
+        :param exclude_tables: Optional list of table names that should not be deleted.
+        """
+        exclude_tables = exclude_tables or []
+
+        with self._connection_context() as connection:
+            if inspect(connection).has_table(self.alembic_version_tbl_name):
+
+                metadata = MetaData()
+                metadata.reflect(bind=connection)
+
+                for schema_table in reversed(metadata.sorted_tables):
```

Comment:
is there a particular reason for going in reverse order? if so, can you add a small comment here?

Reply:
you have to delete them by order of relationships